### PR TITLE
Engine v2

### PR DESCRIPTION
**Pull request by RIg410**

### CODE REVIEW

File: types/src/p2p.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+impl SlotPayload {
+    pub fn size(&self) -> usize {
+        size_of::<Slot>()
+            + size_of::<Hash>() * 3
+            + size_of::<Option<UnixTimestamp>>()
+            + size_of::<Option<u64>>()
+            + self.txs.iter().map(|tx| tx.len()).sum::<usize>()
+    }
```
Comment (i1i1) [inline types/src/p2p.rs]: Lets just use
```suggestion
    pub fn size(&self) -> usize {
        borsh::serialize(self).unwrap().len()
    }
```

File: types/src/p2p.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    pub payload: SlotPayload,
+    pub events: Vec<L2Event>,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct SlotAttribute {
+    slot_id: Slot,
+    events: Vec<L1Event>,
+    sync_status: Option<(Slot, PayloadStatus)>,
+}
```
Comment (i1i1) [inline types/src/p2p.rs:21]: So you've squashed sync status and updates from L1 right? Makes sense so far

Should we make `sync_status: Vec<(Slot, PayloadStatus)>`?

File: types/src/p2p.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    base64::{Base64, Standard},
+    formats::Unpadded,
+};
+use std::mem::size_of;
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct SlotArtifact {
+    pub payload: SlotPayload,
+    pub events: Vec<L2Event>,
+}
```
Comment (i1i1) [inline types/src/p2p.rs]: Lets rename it, as `artifact` is kinda meaningless word. You just get an idea that some computation has been done
```suggestion
pub struct SlotPayloadWithEvents {
    pub payload: SlotPayload,
    pub events: Vec<L2Event>,
}
```

File: engine/src/sub.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        }
+    }
+
+    pub async fn run(mut self) -> Result<(), Error> {
+        loop {
+            if let Some(slot) = self.ledger.get_slot(self.slot).await? {
+                self.sender.send(slot).await?;
+                self.slot += 1;
+            } else {
+                sleep(Duration::from_millis(200)).await;
```
Comment (i1i1) [inline engine/src/sub.rs]: Should we make it configurable?

File: engine/src/lib.rs [hunk 5]
```diff
@@ -1,143 +1,49 @@
...
+use std::sync::Arc;
+
+pub async fn spawn_engine_service<L: Ledger + Send + Sync + 'static>(
+    ledger: L,
+    p2p_cfg: Config,
+    engine_type: EngineType,
+) -> Result<()> {
+    let ledger = Arc::new(ledger);
+    let (mut node, runner) = Node::new(p2p_cfg)?;
+    tokio::spawn(runner.run());
```
Comment (i1i1) [inline engine/src/lib.rs]: Lets move this part into the op-sol or op-move itself

File: types/src/p2p.rs (hunk 1 above)
Comment (RIg410) [reply types/src/p2p.rs]: we basically don't need this function now. So I'll delete it.
but I find the choice to pay through borscht strange.

File: types/src/p2p.rs (hunk 2 above)
Comment (RIg410) [reply types/src/p2p.rs:21]: Slots are finalized sequentially, so I see the point in sending the last currently finalized slot

File: engine/src/lib.rs (hunk 5 above)
Comment (RIg410) [reply engine/src/lib.rs]: make sense

File: types/src/p2p.rs (hunk 1 above)
Comment (i1i1) [reply types/src/p2p.rs]: Sounds good. I generally don't like calculating size with handwritten code.

I'm also not sure if `Vec<Vec<u8>>` size should be equal to the number of `u8`s in it. Probably you should add smth like length for each of the vector

File: types/src/p2p.rs (hunk 2 above)
Comment (i1i1) [reply types/src/p2p.rs:21]: Yes, but the relation between L1 slots and L2 slots are one-to-many, so I just thought that it makes sense to finalize multiple L2 slots at once. WDUT?

Comment (RIg410) [reply types/src/p2p.rs:21]: to finalize several slots at the same time, you simply send the number of the last slot number, and l2 finalizes everything before it

Comment (RIg410) [reply types/src/p2p.rs:21]: There cannot be a case where you have finalized slot 10 but not finalized slot 9

**Review by i1i1 — APPROVED** [review]

**Pull request merged**